### ISSUE TITLE
Tracking issue for `#![feature(async_closure)]` (RFC 2394 / RFC 3668)

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#62290]
This is a tracking issue for `#![feature(async_closure)]` (rust-lang/rfcs#2394).
The feature gate provides the `async |...| expr` closure syntax.

**As with all tracking issues for the language, please file anything unrelated to implementation history, that is: bugs and design questions, as separate issues as opposed to leaving comments here. The status of the feature should also be covered by the feature gate label. Please do not ask about developments here.**

**Steps:**

- [x] Implement the RFC
- [x] Accept another RFC
  - https://github.com/rust-lang/rfcs/pull/3668
- [x] Implement it:  https://github.com/rust-lang/rust/pull/120361
- [x] Adjust documentation ([see instructions on rustc-guide][doc-guide])
- [x] Stabilization PR ([see instructions on rustc-guide][stabilization-guide])
  - #132706 

[stabilization-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs

## Open questions (RFC 2394)

- [x] Interaction between signature deduction and new solver.
  - https://github.com/rust-lang/rust/pull/121857

## Open questions (RFC 3668)

- [x] Which syntax should we choose for bounds?
  - https://github.com/rust-lang/rust/issues/128129

## Related

- https://github.com/rust-lang/rfcs/pull/3668
- https://github.com/rust-lang/rust/issues/127227
- https://github.com/rust-lang/rust/pull/127827

**Comment by khionu** [comment]
Hi, what's the next step for this issue?

**Comment by gilescope** [comment]
@Centril am bumping into this one quite often in beta... from an ergonomics point of view it would be great to get this stable if there's no outstanding concerns / issues... - is this something we might consider putting 'on-deck', or does it need more settlement time?

**Comment by Nemo157** [comment]
Despite "Implement the RFC" being checked this feature is still largely unimplemented (other than the trivial case of `async move || {}` which is equivalent to `move || async move {}`), and as far as I'm aware also requires more design work before it can even really be implemented.

**Comment by khionu** [comment]
@rustbot modify labels to +AsyncAwait-OnDeck

Closures are a very commonly used language feature, especially in APIs like `itertools`. Enabling such APIs to be async has obvious benefits.

**Comment by Centril** [comment]
As @Nemo157 points out this is largely unimplemented and is entirely secondary to fixing *all* the outstanding bugs with the stable part of async/await that we are shipping in 1.39.0.

(Personally I think async closures should be considered holistically with other types of effect-modified closures. Moreover I think they might be largely redundant atop of the combination of closures with async blocks inside them.)

**Comment by CGQAQ** [comment]
Does this will allow ? on Result which have different Error type in the same async closure? currently, there is no way to done that, which is really frustrating

**Comment by Nemo157** [comment]
@CGQAQ do you have an example that fails now? Some [quick testing](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f0b10306b1897b06600f87cdb4eec13b) in the playground makes it look like inference is able to handle at least simple cases correctly (I definitely could see it being possible that the two layers of generics with inferred outputs might cause some issues for inference, and async closures could potentially improve on that by tying those inferred outputs together somehow).

**Comment by CGQAQ** [comment]
@Nemo157 
```rust
async fn a() -> Result<(), reqwest::Error> {Ok(())}
async fn b() -> Result<(), std::io::Error> {Ok(())}

fn c() {
    async {
        a().await?;
        b().await?;
        Ok(())
    };
}
```
```
error[E0282]: type annotations needed
 --> src/x.rs:6:9
  |
6 |         a().await?;
  |         ^^^^^^^^^^ cannot infer type

error: aborting due to previous error
```
```rust
async fn a() -> Result<(), reqwest::Error> {Ok(())}
async fn b() -> Result<(), std::io::Error> {Ok(())}

fn c() {
    async {
        a().await?;
        b().await?;
        Ok::<(), impl std::error::Error>(())
    };
}
```
```
error[E0562]: `impl Trait` not allowed outside of function and inherent method return types
 --> src/x.rs:8:18
  |
8 |         Ok::<(), impl std::error::Error>(())
  |                  ^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

**Comment by Nemo157** [comment]
@CGQAQ the former _is_ unconstrained, there's no way rustc can know what the error type should be, changing it to be used in a closure returning an async block passed to a function which constrains it [works fine](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f18274ebfcf720b5ec9da45aff65f5d). The latter already [works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f6c962ab0c758c76327a2201b3c10734) after replacing `impl Error` with a concrete error type (as far as I'm aware `impl Trait` in that position is not supported and has not been RFCed). (I also switched both to use `serde_json::Error` instead of `reqwest::Error` as it provides the necessary conversion into `io::Error`).

If you want to continue this it would be best to open a new issue and ping me there (or discord) to avoid notifying subscribers of this issue.

**Comment by cdbattags** [comment]
Any recent news/updates for this feature?

**Comment by Centril** [comment]
The purpose of this issue to track the development by noting relevant PRs and whatnot as they happen. The feature gate label is also available to route you to relevant places. You can assume that if no recent developments have been noted here, or linked, or provided in the issues with the label, that there are no such developments.

**Comment by Pzixel** [comment]
Sorry if I missed it, but what's the motivation under this feature? IMO It just adds a slightly different syntax for existing thing (`async ||` vs `|| async`), introducing confusion with no foreseeable benefit. Or at least this is what I see here.

Could you elaborate on that, please?

**Comment by khionu** [comment]
`|| async { }` is not an async closure, it's a closure that contains an async block.

**Comment by Pzixel** [comment]
Yes, I see it, but what's the difference from the user perspective?

**Comment by Nemo157** [comment]
There may be differences in how lifetimes in the arguments are treated once they’re supported. Also elaborating the return type is not possible with the closure to async block ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5340a05b0a17feba7309a80765b5edd9))

**Comment by Jonathas-Conceicao** [comment]
One nice advantage of having async closure is the type signature for the closure's return. While with `async_closure` we can simply have:
```Rust
async || -> i32 { 313 }
```

Without it it's hard to annotate it's return type, since we have a generic `Future` and it has to be `Sized`:
```Rust
// Doesn't compile as it's not `Sized`
|| -> dyn std::future::Future<Output = i32> { async { 313 } }

// Boxed to make it `Sized`
|| -> Box<dyn std::future::Future<Output = i32>> { Box::new(async { 313 }) }
```

**Comment by kaimast** [comment]
Does this feature introduce an `AsyncFn` trait, similar to `core::ops::Fn`? I cannot find it in the RFC.

**Comment by ibraheemdev** [comment]
@kaimast This should become possible once `unboxed_closures` lands:
```rust
#![feature(unboxed_closures, associated_type_bounds, async_closure)]

pub trait AsyncFn<Args>: Fn<Args, Output: core::future::Future<Output = Self::Value>> {
    type Value;
}

impl<F, R, Args> AsyncFn<Args> for F
where
    F: Fn<Args, Output = R>,
    R: core::future::Future,
{
    type Value = R::Output;
}

pub async fn takes_closure<F: AsyncFn<(i32, i32), Value = i32>>(f: F) -> i32 {
    f(23, 32).await
}

fn main() {
    let double = async move |x: i32, y: i32| -> i32 { x * y };
    takes_closure(double);
}
```

**Comment by peepopoggers** [comment]
What is the progress with async closures? Thanks.

**Comment by OfekShochat** [comment]
When will this become not unstable?

**Comment by RpxdYTX** [comment]
How async closures would be expressed as a function parameter? i feel that
```rust
fn takes_async_closure<F: Future<Output = ()>>(_:  impl Fn() -> F)
```
would be too verbose, something like `AsyncFn`, `AsyncFnMut` and `AsyncFnOnce` would be better.
Also, closures that return futures don't work well when a closure parameter is a reference.

**Comment by hamza1311** [comment]
What's the current status of this issue? According to the description, it says only the stabilization is left. How can one help push this to stabilization?

**Comment by 1216892614** [comment]
If this feature stable, will both of these code work like same:
 
1. without feature

```rust
async fn sayer() {
    let mut hs = Vec::new();
    for n in (0..5).rev() {
        hs.push(tokio::spawn(sleep_n(n)));
    }
    for h in hs {
        h.await.unwrap();
    }
}

```

2. with feature

```rust
async fn sayer() {
    let mut hs = Vec::new();
    for n in (0..5).rev() {
        hs.push(tokio::spawn(sleep_n(n)));
    }
    let _ = hs.into_iter().map(async |i|  i.await.unwrap());
}
```

**Comment by peepopoggers** [comment]
> How async closures would be expressed as a function parameter? i feel that
> 
> ```rust
> fn takes_async_closure<F: Future<Output = ()>>(_:  impl Fn() -> F)
> ```
> 
> would be too verbose, something like `AsyncFn`, `AsyncFnMut` and `AsyncFnOnce` would be better. Also, closures that return futures don't work well when a closure parameter is a reference.

Except from these potential keywords being added for convenience, is there any reason to not add async closures?

**Comment by target-san** [comment]
> Aren't async closures just like normal closures but without the redundant `move` keyword? As in:
> 
> ```rust
> move | _ | async move { /* ... */}
> ```
> 
> vs
> 
> ```rust
> async move | _ | { /* ... */ } 
> ```
> 
> is there any other difference other than this?

The first thing is a closure which returns parameterless async block.
The second one is the actual async closure.

**Comment by IndigoLily** [comment]
As @Nemo157 said in [their comment](https://github.com/rust-lang/rust/issues/62290#issuecomment-538782482)

> Despite "Implement the RFC" being checked this feature is still largely unimplemented

Why is this still incorrectly marked as implemented, 4 years later?

**Comment by mzedeler** [comment]
I am trying to create a helper for tests that are supposed to be run with tokios runtime, so the helper has to be async.

What I want is a function that takes a callback as parameter where the tests can executed, something like:

```
#[tokio::test]
async fn non_existing_resource() {
    Model::temp(async |model| {
        ...run tests with model here...
    }) // teardown happens in Model::temp()
}
```

Is there an obvious other way to acheive this?

**Comment by Fancyflame** [comment]
If I understand correctly, just replace `async || {...}` with `|| async {...}`. It acts no differently to async fn.

**Comment by sunrosa** [comment]
So... is this implemented and usable in unstable, or not? I'm seeing it's marked as such, but comments saying otherwise. In what ways is this not implemented?

Also, what are the limitations of the [LendingFn](https://rust-lang.github.io/async-fundamentals-initiative/roadmap/async_closures.html) solution seen here? It seems like a solid solution, or in the very least, like something that should already exist. But because closures have to implement the trait, the compiler needs to have knowledge of it.

**Comment by joebnb** [comment]
4years

**Comment by Fancyflame** [comment]
@kingwingfly Yes, but it's none of the syntax's business, that is because the compiler cannot infer lifetimes inside closure when involving HRTB. And there is a feature solved this problem (I forget the name), I think `async_closure` just simply uses a part of it. This HRTB problem has been disturbing me for a long time.

Let's do some simplification. Whatever marks you add, the following code cannot pass the compilation.
```rust
let _ = |x: &i32| { move || *x };
```

**Comment by c-git** [comment]
Not sure if this is a suitable place to point to the workaround I use. But this is where the compiler points me every time I bump into this. This [post](https://users.rust-lang.org/t/async-closures-and-workarounds/61517) pointed me in the right direction. My main takeaway is put the async on the right of || not the left and then progress becomes possible again (at least for my use cases).

Copy of code so others don't need to follow the link

```rust
async fn with_async_callback<Fut: Future<Output = i32>, F: FnOnce() -> Fut>(f: F) -> i32 {
    f().await
}

with_async_closure(|| async { 123 })
```

**Comment by compiler-errors** [comment]
this is done

**Comment by tisonkun** [comment]
[Just a data point]

Today I meet an issue that when the return type `CallRefFuture` cannot be used in stable channel, some code cannot be written:

```rust
fn schedule_with_arbitrary_delay<F, S, D>(
    mut self,
    mut is_shutdown: F,
    spawn: &S,
    make_delay: D,
    initial_delay: Option<Duration>,
) where
    Self: Sized,
    F: AsyncFnMut() -> () + Send + 'static,
    S: Spawn,
    D: MakeDelay + Send + 'static,
{
    spawn.spawn(async move { .. })
}
```

returns:

```
error[E0277]: `<F as AsyncFnMut<()>>::CallRefFuture<'_>` cannot be sent between threads safely
  --> fastimer/src/schedule/arbitrary.rs:55:15
   |
55 |         spawn.spawn(async move {
   |               ^^^^^ ---------- within this `{async block@fastimer/src/schedule/arbitrary.rs:55:21: 55:31}`
   |               |
   |               `<F as AsyncFnMut<()>>::CallRefFuture<'_>` cannot be sent between threads safely
   |
   = help: within `{async block@fastimer/src/schedule/arbitrary.rs:55:21: 55:31}`, the trait `Send` is not implemented for `<F as AsyncFnMut<()>>::CallRefFuture<'_>`
note: required because it's used within this `async` fn body
  --> fastimer/src/schedule/mod.rs:40:1
   |
40 | / {
41 | |     match select(is_shutdown, f).await {
42 | |         Either::Left(()) => ControlFlow::Break(()),
43 | |         Either::Right(o) => ControlFlow::Continue(o),
44 | |     }
45 | | }
   | |_^
note: required because it's used within this `async` block
  --> fastimer/src/schedule/arbitrary.rs:55:21
   |
55 |         spawn.spawn(async move {
   |                     ^^^^^^^^^^
note: required by a bound in `Spawn::spawn`
  --> fastimer/src/lib.rs:92:39
   |
92 |     fn spawn<F: Future<Output = ()> + Send + 'static>(&self, future: F);
   |                                       ^^^^ required by this bound in `Spawn::spawn`
   = note: the full name for the type has been written to '/Users/tison/Brittani/fastimer/target/debug/deps/fastimer-fa19318f90c9a3a8.long-type-3363693415664072930.txt'
   = note: consider using `--verbose` to print the full type name to the console
help: consider further restricting the associated type
   |
53 |         D: MakeDelay + Send + 'static, <F as AsyncFnMut<()>>::CallRefFuture<'_>: Send
   |                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```


```rust
fn schedule_with_arbitrary_delay<F, S, D>(
    mut self,
    mut is_shutdown: F,
    spawn: &S,
    make_delay: D,
    initial_delay: Option<Duration>,
) where
    Self: Sized,
    F: AsyncFnMut() -> () + Send + 'static,
    for<'a> F::CallRefFuture<'a>: Send,
    S: Spawn,
    D: MakeDelay + Send + 'static,
{
    spawn.spawn(async move { .. })
}
```

returns:

```
error[E0658]: use of unstable library feature `async_fn_traits`
  --> fastimer/src/schedule/arbitrary.rs:52:17
   |
52 |         for<'a> F::CallRefFuture<'a>: Send,
   |                 ^^^^^^^^^^^^^^^^^^^^
```
